### ISSUE TITLE
GraphQL Integration

### ISSUE THREAD

**Issue by hikalkan** [abpframework/abp#86]
### Intro

- [ ] Pick a C# GraphQL library. 

* see https://github.com/graphql-dotnet/graphql-dotnet and https://github.com/ChilliCream/hotchocolate (https://github.com/volosoft/abp/issues/340)

### Modules

Create GraphQL endpoints for all modules.

- [ ] Identity
- [ ] Tenant management

### Resources

* https://fullstackmark.com/post/17/building-a-graphql-api-with-aspnet-core-2-and-entity-framework-core

**Comment by hitaspdotnet** [comment]
@hikalkan Hi.
Is it must be a part depended on Abp.Ddd.Application? Can you suggest me for package naming? And where is the best place for this package?
```
Volo.Abp.GraphQL
-GraphQL (original net core library)
-Volo
--Abp
---GraphQL
Volo.Abp.GraphQL.GraphiQL
-Volo
--Abp
---GraphiQL
```
Can I use this solution map?

**Comment by hikalkan** [comment]
Actually, I haven't worked much about how ABP can integrate to GraphQL. I don't know yet what Volo.Abp.GraphQL package will provide on top of the GraphQL library. But I believe we can provide some base classes and services to create GraphQL endpoints easier.

Do you have an idea? Can it be integrated to `IRepository<TEntity, TKey>` for instance?

>Is it must be a part depended on Abp.Ddd.Application

Not sure about that.

I suggest you to start with the implementation in your mind and we see what you are trying to do :) For now, I will accept what you contribute. I suggest you to contribute it progressively, a basic implementation at first.

**Comment by hitaspdotnet** [comment]
@hikalkan 

For now I want to add GraphQL as a ABP MiddleWare, to add the GraphQL implementation as a different part of the Application. Like this we're able to use REST based POST and PUT request to add or update the data and to use the GraphQL to query the data.

I also want that the middleware is listening to the own path like old version of ABP.

>  Can it be integrated to IRepository<TEntity, TKey> for instance?

You need to think about what data you wanna provide and you need to think about how you wanna provide that data then you can build your own repositories for query to QL server.
Using IRepository<TEntity, TKey> for building queries not answered in my ZERO project because I haven't good idea about ABP basic repository but i'm sure ABP has created by great team. You can make everything possible.

**Comment by hikalkan** [comment]
>we're able to use REST based POST and PUT request to add or update the data and to use the GraphQL to query the data

I agree on that. Every module has its own REST APIs. In addition, every module will provide a GraphQL API to provide more elastic and powerful querying.

**Comment by hitaspdotnet** [comment]
>  Pick a C# GraphQL library.

@hikalkan Hi,

At the moment the [GraphQL-Server library](https://github.com/graphql-dotnet/server) and [GraphQL-Conventions](https://github.com/graphql-dotnet/conventions) are the best on the web. 

The conventions library allows you to automatically wrap your .NET classes into GraphQL schema definitions using existing property getters and methods as field resolvers. Contains attribute based operation registering, that good for easy to use in application level but needs some customize. We can building a module top of this library.

Server library as you can guess from the library name, this package build Server + WebSockets Transport. If you want to have a html based GraphQL UI of all GraphQL's types (contains all modules schema) in application you need that.

**Comment by leonkosak** [comment]
See also [graphql-net](https://github.com/chkimes/graphql-net) and [GraphQL.EntityFramework](https://github.com/SimonCropp/GraphQL.EntityFramework) for converting GraphQL queries to EF Core IQueryable.

Take a look at [graphql-to-mongodb](https://github.com/Soluto/graphql-to-mongodb) library for integration with MongoDB.

Those are not official libraries (provided by Microsoft and MongoDB Inc.), but efficient translations from GraphQL queries to IQueryable/Mongo query could be one of the biggest "missing links" when using GraphQL APIs.

**Comment by Trojaner** [comment]
I think the best solution would be to write a GraphQL gateway which proxies requests to each module's REST API. Keep in mind that one of the benefits of GraphQL is reduced RTT as you only need a single request. If each module defines it's own GraphQL endpoint it would defeat the purpose of GraphQL. 

It would need an efficient way to describe how entities and their APIs are related. Assume I have two entities called Order
 and Product defined in separate modules. I should be still able to request an order with all of it's products in a single request. This gateway would have to do it similar to how BFFs do it.

Another issue is that while GraphQL is flexible when it comes to breaking changes, REST APIs are usually not.

**Comment by Loongle** [comment]
Is there any update

**Comment by leonkosak** [comment]
I think that currently gRPC is more important to Volosoft than GraphQL and therefore no priority before abp v4.0.
Blazor UI option and support for .NET 5 are top goals for abp v4.0).

**Comment by hikalkan** [comment]
@leonkosak writes a good answer, thanks :)

**Comment by hikalkan** [comment]
BTW, you can always create GraphQL endpoints for your own application, no restriction.

**Comment by abhishekbhalani** [comment]
> @hikalkan
> 
> For now I want to add GraphQL as a ABP MiddleWare, to add the GraphQL implementation as a different part of the Application. Like this we're able to use REST based POST and PUT request to add or update the data and to use the GraphQL to query the data.
> 
> I also want that the middleware is listening to the own path like old version of ABP.
> 
> > Can it be integrated to IRepository<TEntity, TKey> for instance?
> 
> You need to think about what data you wanna provide and you need to think about how you wanna provide that data then you can build your own repositories for query to QL server.
> Using IRepository<TEntity, TKey> for building queries not answered in my ZERO project because I haven't good idea about ABP basic repository but i'm sure ABP has created by great team. You can make everything possible.

@hitaspdotnet Good idea.
I did not worked with graphql more. 
Share me if you get integration with abp+mongodb app structure..

**Comment by Loongle** [comment]
Is there any update? 😊

**Comment by tvddev** [comment]
Any further work planned on this?

**Comment by gdlcf88** [comment]
Hey guys, I have published an ABP module for GraphQL integration that implemented with [graphql-dotnet](https://github.com/graphql-dotnet/graphql-dotnet):

https://github.com/EasyAbp/Abp.GraphQL.

I made an [interesting schema](https://github.com/EasyAbp/Abp.GraphQL/blob/main/src/EasyAbp.Abp.GraphQL.Provider.GraphQLDotnet/EasyAbp/Abp/GraphQL/Provider/GraphQLDotnet/AppServiceSchema.cs) that provides GraphQL query for any **IReadOnlyAppService**.

Normally, we use RESTful API to get an entity or list of entities:

```http request
GET /api/book/CA2EBE5D-D0DC-4D63-A77A-46FF520AEC44
```
```http request
GET /api/book
```

But now, we can get them in the GraphQL way:

![image](https://user-images.githubusercontent.com/30018771/147388601-21a19507-d39b-4c6e-a066-883152aa51ac.png)

All you need to do is install and configure the GraphQL module like:

```c#
Configure<AbpGraphQLOptions>(options =>
{
    // Find entities: Book, Author, City...
    options.AppServiceSchemes.Configure(
        typeof(MyProjectApplicationContractsModule).Assembly);

    // Find entities: IdentityUser, IdentityRole
    options.AppServiceSchemes.Configure(
        typeof(AbpIdentityApplicationContractsModule).Assembly);
});
```

Custom schemas are supported by the dependency injection:

```c#
public class MyCustomSchema : Schema, ITransientDependency
{
   public MySchema(IServiceProvider serviceProvider) : base(serviceProvider)
   {
       Query = serviceProvider.GetRequiredService<MyCustomQuery>();
   }
}
```

The mutation and subscription features are not supported yet, but I believe they are not difficult to implement.

I'm not really sure if this module is enough to meet your needs, suggestions are welcome.
